### PR TITLE
An alternative to asyncio.wait and asyncio.gather function with correct cancellation

### PR DESCRIPTION
**Pull request by mosquito**

### CODE REVIEW

File: tests/test_utils.py [hunk 1]
```diff
@@ -6,7 +7,8 @@
 
 import pytest
 
-from aiomisc.utils import chunk_list, bind_socket
+from aiomisc.entrypoint import entrypoint
+from aiomisc.utils import chunk_list, bind_socket, wait_for
```
Comment (alvassin) [inline tests/test_utils.py]: Imports not sorted! 😡

File: aiomisc/utils.py [hunk 2]
```diff
@@ -73,3 +73,84 @@ def new_event_loop(pool_size=None) -> asyncio.AbstractEventLoop:
...
+def wait_for(*coros, raise_first: bool = True,
+             loop: asyncio.AbstractEventLoop = None):
+
+    tasks = list()
+    loop = loop or asyncio.get_event_loop()
+
+    result = loop.create_future()
+    waiting = len(coros)
+
+    def cancelling():
```
Comment (alvassin) [inline aiomisc/utils.py]: я бы назвал бы cancel_undone()

Comment (mosquito) [reply aiomisc/utils.py]: It's calling in finally.

File: aiomisc/utils.py [hunk 3]
```diff
@@ -73,3 +73,84 @@ def new_event_loop(pool_size=None) -> asyncio.AbstractEventLoop:
     asyncio.set_event_loop(loop)
 
     return loop
+
+
+def wait_for(*coros, raise_first: bool = True,
+             loop: asyncio.AbstractEventLoop = None):
+
+    tasks = list()
```
Comment (leenr) [inline aiomisc/utils.py]: Better `type: List[asyncio.Task]` here instead of `type: asyncio.Task` in `for`.

File: aiomisc/utils.py [hunk 4]
```diff
@@ -73,3 +73,84 @@ def new_event_loop(pool_size=None) -> asyncio.AbstractEventLoop:
...
+    loop = loop or asyncio.get_event_loop()
+
+    result = loop.create_future()
+    waiting = len(coros)
+
+    def cancel_undone():
```
Comment (leenr) [inline aiomisc/utils.py]: `cancel_pending` or `cancel_pending`.

File: aiomisc/utils.py [hunk 5]
```diff
@@ -73,3 +73,85 @@ def new_event_loop(pool_size=None) -> asyncio.AbstractEventLoop:
     asyncio.set_event_loop(loop)
 
     return loop
+
+
+def wait_for(*coros, raise_first: bool = True, cancel: bool = True,
```
Comment (leenr) [inline aiomisc/utils.py]: `cancel_on_exception` will be more readable.

File: aiomisc/utils.py [hunk 6]
```diff
@@ -73,3 +73,85 @@ def new_event_loop(pool_size=None) -> asyncio.AbstractEventLoop:
...
+             loop: asyncio.AbstractEventLoop = None):
+
+    tasks = list()
+    loop = loop or asyncio.get_event_loop()
+
+    result = loop.create_future()
```
Comment (leenr) [inline aiomisc/utils.py]: `result_future` or just `future`?
E.g:
`future.cancel()` vs `result.set_result(results)`
`return await future` vs `return await result`

File: aiomisc/utils.py [hunk 7]
```diff
@@ -73,3 +73,85 @@ def new_event_loop(pool_size=None) -> asyncio.AbstractEventLoop:
     asyncio.set_event_loop(loop)
 
     return loop
+
+
+def wait_for(*coros, raise_first: bool = True, cancel: bool = True,
+             loop: asyncio.AbstractEventLoop = None):
+
+    tasks = list()
```
Comment (leenr) [inline aiomisc/utils.py]: Better `type: List[asyncio.Task]` here instead of `type: asyncio.Task` in `for` below.

File: aiomisc/utils.py [hunk 8]
```diff
@@ -73,3 +73,89 @@ def new_event_loop(pool_size=None) -> asyncio.AbstractEventLoop:
     asyncio.set_event_loop(loop)
 
     return loop
+
+
+_TASKS_LIST = List[asyncio.Task]
```
Comment (leenr) [inline aiomisc/utils.py]: That constant for what?

File: aiomisc/utils.py [hunk 9]
```diff
@@ -3,7 +3,7 @@
 import logging.handlers
 import socket
 from multiprocessing import cpu_count
-from typing import Iterable, Any, Tuple
+from typing import Iterable, Any, Tuple, Coroutine, List
```
Comment (leenr) [inline aiomisc/utils.py]: Not sorted :angry:

File: aiomisc/utils.py (hunk 8 above)
Comment (mosquito) [reply aiomisc/utils.py]: Pylama will be broken otherwise.

**Pull request merged**